uplink/storage/segments/store.go

```diff
@@ -68,133 +66,103 @@ func NewSegmentStore(metainfo *metainfo.Client, ec ecclient.Client, rs eestream.
 	remoteSized, err := peekReader.IsLargerThan(s.thresholdSize)
 	if err != nil {
 		return Meta{}, err
 	}
 
-	var path storj.Path
-	var pointer *pb.Pointer
-	var originalLimits []*pb.OrderLimit
 	if !remoteSized {
-		p, metadata, err := segmentInfo()
+		segmentIndex, encryption, err := segmentInfo()
 		if err != nil {
 			return Meta{}, Error.Wrap(err)
 		}
-		path = p
-
-		pointer = &pb.Pointer{
-			CreationDate:   time.Now(),
-			Type:           pb.Pointer_INLINE,
-			InlineSegment:  peekReader.thresholdBuf,
-			SegmentSize:    int64(len(peekReader.thresholdBuf)),
-			ExpirationDate: expiration,
-			Metadata:       metadata,
-		}
-	} else {
-		// early call to get bucket name, rest of the path cannot be determine yet
-		p, _, err := segmentInfo()
-		if err != nil {
-			return Meta{}, Error.Wrap(err)
-		}
-		bucket, objectPath, _, err := splitPathFragments(p)
-		if err != nil {
-			return Meta{}, err
-		}
 
-		// path and segment index are not known at this point
-		limits, rootPieceID, piecePrivateKey, err := s.metainfo.CreateSegment(ctx, bucket, objectPath, -1, redundancy, s.maxEncryptedSegmentSize, expiration)
-		if err != nil {
-			return Meta{}, Error.Wrap(err)
-		}
-
-		sizedReader := SizeReader(peekReader)
-
-		successfulNodes, successfulHashes, err := s.ec.Put(ctx, limits, piecePrivateKey, s.rs, sizedReader, expiration)
+		err = s.metainfo.MakeInlineSegment(ctx, metainfo.MakeInlineSegmentParams{
+			StreamID: streamID,
+			Position: storj.SegmentPosition{
+				Index: int32(segmentIndex),
+			},
+			Encryption:          encryption,
+			EncryptedInlineData: peekReader.thresholdBuf,
+		})
 		if err != nil {
 			return Meta{}, Error.Wrap(err)
 		}
+		return Meta{}, nil
+	}
 
-		p, metadata, err := segmentInfo()
-		if err != nil {
-			return Meta{}, Error.Wrap(err)
-		}
-		path = p
+	segmentID, limits, piecePrivateKey, err := s.metainfo.BeginSegment(ctx, metainfo.BeginSegmentParams{
+		StreamID:     streamID,
+		MaxOderLimit: s.maxEncryptedSegmentSize,
+	})
+	if err != nil {
+		return Meta{}, Error.Wrap(err)
+	}
 
-		pointer, err = makeRemotePointer(successfulNodes, successfulHashes, s.rs, rootPieceID, sizedReader.Size(), expiration, metadata)
-		if err != nil {
-			return Meta{}, Error.Wrap(err)
-		}
+	sizedReader := SizeReader(peekReader)
 
-		originalLimits = make([]*pb.OrderLimit, len(limits))
-		for i, addressedLimit := range limits {
-			originalLimits[i] = addressedLimit.GetLimit()
-		}
+	successfulNodes, successfulHashes, err := s.ec.Put(ctx, limits, piecePrivateKey, s.rs, sizedReader, expiration)
+	if err != nil {
+		return Meta{}, Error.Wrap(err)
 	}
 
-	bucket, objectPath, segmentIndex, err := splitPathFragments(path)
+	segmentIndex, encryption, err := segmentInfo()
 	if err != nil {
-		return Meta{}, err
+		return Meta{}, Error.Wrap(err)
 	}
 
-	savedPointer, err := s.metainfo.CommitSegment(ctx, bucket, objectPath, segmentIndex, pointer, originalLimits)
+	uploadResults := make([]*pb.SegmentPieceUploadResult, 0)
```

Comment:
let's initialize the slice with the capacity `make([]*pb.SegmentPieceUploadResult, 0, len(successfulNode))`